### PR TITLE
Handle deploy for ERC20

### PR DESCRIPTION
**Pull request by tcharding**
Add support for the deploy action.

Currently we do not take the deployment of the ERC20 contract into account. Do so by adding variants to the `AliceState` and `BobState` enums and conditional logic depending on the protocol in use when determining next action and next state.

Note: Re-introduces a `Protocol` type. Do we want this?

- Patch 1 is refactor: Add `use` statement to reduce noise.
- Patch 2 is the ERC20 support.
- Patch 3 is refactor: Use 'broadcast' uniformly throughout the module.
- Patch 4 is whitespace: audit TODOs and tidy up the TODO comments.
- Patch 5 adds units test for complete swaps in all directions.

### CODE REVIEW

File: comit/src/expiries.rs [hunk 1]
```diff
@@ -908,4 +1030,64 @@ mod tests {
 
         assert_that!(got_action).is_equal_to(want_action);
     }
+
+    #[tokio::test]
+    async fn alice_next_action_after_start_is_deploy_herc20_hbit() {
+        let start_at = Timestamp::now();
+        let (ac, bc) = mock_connectors();
```
Comment (thomaseizinger) [inline comit/src/expiries.rs:1077]: :guitar: 

```suggestion
        let (ac, dc) = mock_connectors();
```

**Review by thomaseizinger — APPROVED** [review]
I didn't audit the implementation much but the tests look ok :)

File: comit/src/expiries.rs (hunk 1 above)
Comment (thomaseizinger) [reply comit/src/expiries.rs:1148]: Wait, I've got an even better one!

**Review by D4nte — APPROVED** [review]

File: comit/src/expiries.rs [hunk 2]
```diff
@@ -6,14 +6,18 @@
...
 use num::integer;
 use std::{cmp, fmt};
 use time::Duration;
 
-// TODO: Currently we ignore deploy for Erc20.
+#[cfg(not(test))]
+use tracing::debug;
+
+#[cfg(test)]
+use std::println as debug;
```
[the remark this answers is not in the document]
Comment (tcharding) [reply comit/src/expiries.rs:20]: Oh, that's not supposed to be still in there. I add that any time I'm debugging, sweet huh.

File: comit/src/expiries.rs [hunk 3]
```diff
@@ -287,7 +306,10 @@ fn period_for_alice_to_complete(config: &Config, current_state: AliceState) -> D
             return acc;
         }
 
-        let (_action, next_state) = state.next();
+        let (_action, next_state) = match config.protocol() {
```
[the remark this answers is not in the document]
Comment (tcharding) [reply comit/src/expiries.rs:330]: Cheers, I'l have a play with it.

File: comit/src/expiries.rs [hunk 4]
```diff
@@ -433,13 +469,34 @@ impl AliceState {
...
+    /// Gets the next action required to transition to the next state for a
+    /// hbit-herc20 swap.
+    fn next_hbit_herc20(&self) -> (AliceAction, AliceState) {
         use self::{AliceAction::*, AliceState::*};
 
         match self {
             None => (Start, Started),
             Started => (FundAlpha, FundAlphaTransactionBroadcast),
+            DeployAlphaTransactionBroadcast => unreachable!("hbit-herc20 no deploy for Alice"),
+            AlphaDeployed => unreachable!("hbit-herc20 no deploy for Alice"),
```
[the remark this answers is not in the document]
Comment (tcharding) [reply comit/src/expiries.rs:520]: Yes, of course.

File: comit/src/expiries.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply comit/src/expiries.rs:20]: Damn, sloppy work by me.

File: comit/src/expiries.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply comit/src/expiries.rs:330]: Had a look, I'm not really sure what you mean. I'm not totally happy with the `Config` handling throughout this module - open to suggestions.

File: comit/src/expiries.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (luckysori) [reply comit/src/expiries.rs]: Is `dbg!` not good enough?

File: comit/src/expiries.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply comit/src/expiries.rs:309]: I think I underestimated the importance of the context of the call site on the returned variable.

**Pull request merged**